### ISSUE TITLE
Tables and merging?

### ISSUE THREAD

**Issue by tmccombs** [mehcode/config-rs#4]
Consider a configuration with a default of "a.a", one source that provides "a.b" and "a.c", a second source that provides "a.c" and "a.d" and an override for "a.e". If you get any of these specific keys from the config, it works fine. However, if you just get "a", then you will get a table that only has the "c" and "d" keys from the second source.  It seems like there should be a way to get a merged object with all five keys.

One way this could be done would be to have a `SubConfig` object with a reference to the root config and a prefix that it uses to do lookups.

**Comment by tmccombs** [comment]
Actually, on closer inspection it looks like the implementation doesn't actually support getting tables or arrays (at least for json or toml).

**Comment by mehcode** [comment]
@tmccombs Yeah. The v0.1 implementation just doesn't support it at all. 

But funny you should make this issue. My v0.2 task list is:

```
 ☐ `File::from_str` — Abstract _where_ the file comes from
 ☐ Support `Table` — `config::get_map`
 ☐ Support `Array` — `config::get_slice`
 ☐ Support serde to deserialize configuration into a deserialize-able
```

**Comment by mehcode** [comment]
@tmccombs Can you test against master? I just pushed table/array support. This includes the deep merging you're talking about.

**Comment by tmccombs** [comment]
Awesome!

I'll close this since the issue addressed is solved. 

I do have a couple of concerns, although they certainly don't need to be addressed right away:

1. It looks like using a dot-separated key to get values no longer works (although setting with a dot seperated path does), your suggestion of using a subset of JSON Path sounds good to me 
2. Performing the deep merge when configs are merged, rather than when config is requested means that dynamic Sources (for example etcd, a file that watches for changes with inotify, or possibly environment variables) must be able to enumerate all available config when merged and somehow notify the config when values change. None of the current sources currently require this, but it is something to keep in mind.

**Comment by mehcode** [comment]
> Performing the deep merge when configs are merged, rather than when config is requested means that dynamic Sources [...]

I did some thinking and decided to optimize for configuration retrieval instead of insertion. I am still storing the configuration stack so it can be re-built with the same priority if a source updates. That's the plan anyway.

---

> It looks like using a dot-separated key to get values no longer works (although setting with a dot seperated path does), your suggestion of using a subset of JSON Path sounds good to me

Yeah. I didn't like the implementation of that in the first release so I trashed it when I was moving things around. I'll at least bring back simple `a.b.c` retrieval before I make the next release. 

https://github.com/mehcode/config-rs/issues/5
